dace/distr_types.py

```diff
@@ -0,0 +1,600 @@
+    a matrix to a 2D process-grid, but tile the matrix rows over the grid's columns, then `correspondence = [1, 0]`.
+    """
+
+    name = Property(dtype=str, desc="The type's name.")
+    dtype = TypeClassProperty(default=dtypes.int32, choices=dtypes.Typeclasses)
+    shape = ShapeProperty(default=[], desc="The array's shape.")
+    subshape = ShapeProperty(default=[], desc="The sub-array's shape.")
+    pgrid = Property(dtype=str,
+                     allow_none=True,
+                     default=None,
+                     desc="Name of the process-grid where the data are distributed.")
+    correspondence = ListProperty(int,
+                                  allow_none=True,
+                                  default=None,
+                                  desc="Correspondence of the array's indices to the process grid's "
+                                  "indices.")
+
+    def __init__(self,
+                 name: str,
+                 dtype: dtypes.typeclass,
+                 shape: ShapeType,
+                 subshape: ShapeType,
+                 pgrid: str = None,
+                 correspondence: Sequence[Integral] = None):
+        self.name = name
+        self.dtype = dtype
+        self.shape = shape
+        self.subshape = subshape
+        self.pgrid = pgrid
+        self.correspondence = correspondence or list(range(len(shape)))
+        self._validate()
+
+    def validate(self):
+        """ Validate the correctness of this object.
+            Raises an exception on error. """
+        self._validate()
+
+    # Validation of this class is in a separate function, so that this
+    # class can call `_validate()` without calling the subclasses'
+    # `validate` function.
+    def _validate(self):
+        if any(not isinstance(s, (Integral, symbolic.SymExpr, symbolic.symbol, symbolic.sympy.Basic))
+               for s in self.shape):
+            raise TypeError('Shape must be a list or tuple of integer values or symbols')
+        if any(not isinstance(s, (Integral, symbolic.SymExpr, symbolic.symbol, symbolic.sympy.Basic))
+               for s in self.subshape):
+            raise TypeError('Sub-shape must be a list or tuple of integer values or symbols')
+        if any(not isinstance(i, Integral) for i in self.correspondence):
+            raise TypeError('Correspondence must be a list or tuple of integer values')
+        if len(self.shape) != len(self.subshape):
+            raise ValueError('The dimensionality of the shape and sub-shape must match')
+        if len(self.correspondence) != len(self.shape):
+            raise ValueError('The dimensionality of the shape and correspondence list must match')
+        return True
+
+    def to_json(self):
+        attrs = serialize.all_properties_to_json(self)
+        retdict = {"type": type(self).__name__, "attributes": attrs}
+        return retdict
+
+    @classmethod
+    def from_json(cls, json_obj, context=None):
```

Comment:
looks like this function and the above one could benefit from the half-working `__new__` fix.

Reply:
I followed the same class structure/methods as the descriptors in `dace/data`. Where can I find an example of this `__new__` fix?

Reply:
in the failing PR :D